be/src/exec/vectorized/json_scanner.cpp

```diff
@@ -272,56 +278,133 @@ Status JsonReader::close() {
+            // Expand array.
+            auto arr = doc.get_array().value();
+            if (_scanner->_json_paths.empty()) {
+                RETURN_IF_ERROR(_process_array(chunk, slot_descs, arr));
+            } else {
+                RETURN_IF_ERROR(_process_array_with_json_path(chunk, slot_descs, arr));
             }
+            break;
+        }
+
+        case simdjson::ondemand::json_type::object: {
+            auto obj = doc.get_object().value();
             if (_scanner->_json_paths.empty()) {
-                for (SlotDescriptor* slot_desc : slot_descs) {
-                    if (slot_desc == nullptr) {
-                        continue;
-                    }
-                    ColumnPtr& column = chunk->get_column_by_slot_id(slot_desc->id());
-                    const char* column_name = slot_desc->col_name().c_str();
-                    if (!objectValue->IsObject() || !objectValue->HasMember(column_name)) {
-                        column->append_nulls(1);
-                    } else {
-                        _construct_column((*objectValue)[column_name], column.get(), slot_desc->type());
-                    }
-                }
+                RETURN_IF_ERROR(_process_object(chunk, slot_descs, obj));
             } else {
-                size_t slot_size = slot_descs.size();
-                size_t jsonpath_size = _scanner->_json_paths.size();
-                for (size_t i = 0; i < slot_size; i++) {
-                    if (slot_descs[i] == nullptr) {
-                        continue;
-                    }
-                    ColumnPtr& column = chunk->get_column_by_slot_id(slot_descs[i]->id());
-                    if (i >= jsonpath_size) {
-                        column->append_nulls(1);
-                        continue;
-                    }
-                    rapidjson::Value* json_values = JsonFunctions::get_json_object_from_parsed_json(
-                            _scanner->_json_paths[i], objectValue, _origin_json_doc.GetAllocator());
-                    if (json_values == nullptr) {
-                        column->append_nulls(1);
-                    } else {
-                        _construct_column(*json_values, column.get(), slot_descs[i]->type());
-                    }
-                }
+                RETURN_IF_ERROR(_process_object_with_json_path(chunk, slot_descs, obj));
             }
-            rows_to_read--;
-            _next_line++;
+            break;
+        }
+
+        default: {
+            std::string err_msg("JSON data type is not supported");
+            _state->append_error_msg_to_file("", err_msg);
+            _counter->num_rows_filtered++;
+            return Status::DataQualityError(err_msg.c_str());
+        }
+        }
+    }
+    return Status::OK();
+}
+
+Status JsonReader::_process_array(Chunk* chunk, const std::vector<SlotDescriptor*>& slot_descs,
+                                  simdjson::ondemand::array& arr) {
+    for (auto a : arr) {
```

Comment:
auto& a

Reply:
> Array Iteration: To iterate through an array, use for (auto value : array) { ... }. This will step through each value in the JSON array.

It's the recommended way to iterate an array. 
https://github.com/simdjson/simdjson/blob/master/doc/basics.md#json-pointer
Explicitly define `a` as reference would lead error `initial value of reference to non-const must be an lvalue`